### PR TITLE
Keyless Tables Spec

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**
skipped BATS for defining keyless tables semantics

### CODE REVIEW

File: bats/keyless-tables.bats [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+    [[ "$output" = "1,1" ]] || false
+}
+
+@test "delete from keyless" {
+    run dolt sql -q "DELETE FROM keyless WHERE c1 = 2;"
+    [ $status -eq 0 ]
+
+    run dolt sql -q "SELECT * FROM keyless;" -r csv
+    [ $status -eq 0 ]
+    [[ "$output" = "0,0" ]] || false
```
Comment (timsehn) [inline bats/keyless-tables.bats:57]: Lines again since you're testing order

File: bats/keyless-tables.bats [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    [ $status -eq 0 ]
+    run dolt sql -q "SELECT * FROM keyless WHERE c0 > 6;" -r csv
+    [ $status -eq 0 ]
+    [[ "$output" = "7,7" ]] || false
+    [[ "$output" = "7,7" ]] || false
+    [[ "$output" = "8,8" ]] || false
+    [[ "$output" = "8,8" ]] || false
+    [[ "$output" = "9,9" ]] || false
+    [[ "$output" = "9,9" ]] || false
+}
```
Comment (timsehn) [inline bats/keyless-tables.bats:369]: Aaron is worried about deletes. I think we need delete tests. and more mixes of updates, inserts and deletes.

And I think regexes are going to be a problem specifically on tests where you assume order. I would use lines and equalities instead...at least lines.

File: bats/keyless-tables.bats [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+    dolt commit -am "inserted on other"
+
+    run dolt diff master
+    [ $status -eq 0 ]
+    [[ "$output" = "|  +  | 7  | 7  |" ]] || false
+    [[ "$output" = "|  -  | 8  | 8  |" ]] || false
+    [[ "$output" = "|  +  | 8  | 8  |" ]] || false
+    [[ "$output" = "|  -  | 9  | 9  |" ]] || false
+    [[ "$output" = "|  +  | 9  | 9  |" ]] || false
+}
```
Comment (andrew-wm-arthur) [inline bats/keyless-tables.bats:346]: Honestly,  this test makes me want bag semantics

File: bats/keyless-tables.bats [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    [[ "$output" = "3" ]] || false
+    run dolt sql -q "SELECT * FROM keyless;" -r csv
+    [ $status -eq 0 ]
+    [[ "$output" = "7,7" ]] || false
+    [[ "$output" = "8,8" ]] || false
+    [[ "$output" = "9,9" ]] || false
+
+}
+
+@test "keyless diff against working set" {
```
Comment (zachmu) [inline bats/keyless-tables.bats:138]: Would be good to see some additional tests here, like delete * followed by inserting the original values. Does it show a diff? (hopefully not)

Updating one column in the first row. What does the diff have?

File: bats/keyless-tables.bats [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+
+    dolt checkout other
+    dolt sql -q "INSERT INTO keyless VALUES (7,7),(8,8),(9,9);"
+    dolt commit -am "inserted on other"
+
+    run dolt merge master
+    [ $status -eq 0 ]
+    run dolt sql -q "SELECT * FROM keyless WHERE c0 > 6;" -r csv
+    [ $status -eq 0 ]
+    [[ "$output" = "7,7" ]] || false
```
Comment (zachmu) [inline bats/keyless-tables.bats:186]: Is this supposed to be checking that there is only one copy of each row? It's not testing that

File: bats/keyless-tables.bats [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+@test "keyless diff delete+add against working" {
+    dolt sql <<SQL
+DELETE FROM keyless WHERE c0 = 2;
+INSERT INTO keyless VALUES (2,2)
+SQL
+    run dolt diff
+    [ $status -eq 0 ]
+    [[ "$output" = "|  +  | 2  | 2  |" ]] || false
+    [[ "$output" = "|  -  | 2  | 2  |" ]] || false
+}
```
Comment (andrew-wm-arthur) [inline bats/keyless-tables.bats:414]: This one also makes me want bag semantics

File: bats/keyless-tables.bats [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+
+    dolt checkout other
+    dolt sql -q "INSERT INTO keyless VALUES (9,9),(8,8),(7,7),;"
+    dolt commit -am "inserted on other"
+
+    run dolt merge master
+    [ $status -eq 0 ]
+    run dolt sql -q "SELECT * FROM keyless WHERE c0 > 6;" -r csv
+    [ $status -eq 0 ]
+    [[ "$output" = "7,7" ]] || false
```
Comment (zachmu) [inline bats/keyless-tables.bats:280]: Same comment here, you aren't checking line by line in this test

File: bats/keyless-tables.bats [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+@test "keyless diff branches with convergent mutation history" {
+    dolt branch other
+
+    dolt sql -q "INSERT INTO keyless VALUES (7,7),(8,8),(9,9);"
+    dolt commit -am "inserted on master"
+
+    dolt checkout other
+    dolt sql <<SQL
+INSERT INTO keyless VALUES (9,9),(8,8),(7,7),;
+UPDATE keyless SET (c0,c1) = (7,7) WHERE c0 = 9;
```
Comment (zachmu) [inline bats/keyless-tables.bats:296]: Do we support this syntax?

**Review by zachmu — COMMENTED** [review]
Second tim's comments about deletes.

**Pull request closed without merging**